File: iOS/ssabree/ssabree/Features/GroupWrite/GroupWriteViewModel.swift
import Foundation
import Observation

struct GroupWriteUiState {
    var isSubmitting: Bool = false
    var isSuccess: Bool = false
    var errorMessage: String? = nil
    var campusId: Int? = nil
}

@Observable
final class GroupWriteViewModel {
    private let groupRepository: GroupRepository
    private let myPageRepository: MyPageRepository
    private let campusRepository: CampusRepository

    var uiState = GroupWriteUiState()

    init(
        groupRepository: GroupRepository,
        myPageRepository: MyPageRepository,
        campusRepository: CampusRepository
    ) {
        self.groupRepository = groupRepository
        self.myPageRepository = myPageRepository
        self.campusRepository = campusRepository
    }

    @MainActor
    func loadCampusId() async {
        let myPageResult = await myPageRepository.getMyPage()
        guard case .success(let myPage) = myPageResult,
              let campusName = myPage.user?.campus?.trimmingCharacters(in: .whitespaces),
              !campusName.isEmpty else {
            uiState.campusId = nil
            return
        }

        let campusesResult = await campusRepository.getCampuses()
        guard case .success(let campuses) = campusesResult else {
            uiState.campusId = nil
            return
        }

        let normalized = normalizeCampusName(campusName)
        uiState.campusId = campuses.first { campus in
            let campusNormalized = normalizeCampusName(campus.name)
            return campusNormalized == normalized ||
                   campusNormalized.contains(normalized) ||
                   normalized.contains(campusNormalized)
        }?.id
    }

    @MainActor
    func submit(kind: GroupKind, info: GroupCreateInfo) async {
        uiState.isSubmitting = true
        uiState.errorMessage = nil

        let result: Result<GroupSummaryModel, Error>
        switch kind {
        case .study:
            result = await groupRepository.createStudy(info: info)
        case .project:
            result = await groupRepository.createTeam(info: info)
        }

        uiState.isSubmitting = false

        switch result {
        case .success:
            uiState.isSuccess = true
        case .failure(let error):
            uiState.errorMessage = error.localizedDescription
        }
    }

    func resetResult() {
        uiState.isSuccess = false
        uiState.errorMessage = nil
    }

    private func normalizeCampusName(_ name: String) -> String {
        return name.replacingOccurrences(of: "캠퍼스", with: "")
            .replacingOccurrences(of: " ", with: "")
            .lowercased()
    }
}

File: android/ssabree/app/src/main/java/com/ssafy/ssabree/features/groupwrite/screen/GroupWriteViewModel.kt
package com.ssafy.ssabree.features.groupwrite.screen

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ssafy.ssabree.core.repository.CampusRepository
import com.ssafy.ssabree.core.repository.GroupRepository
import com.ssafy.ssabree.core.repository.MemberRepository
import com.ssafy.ssabree.core.repository.model.GroupCreateInfo
import com.ssafy.ssabree.core.repository.model.GroupUpdateInfo
import com.ssafy.ssabree.features.group.model.GroupKind
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class GroupWriteUiState(
    val isSubmitting: Boolean = false,
    val isSuccess: Boolean = false,
    val campusId: Long? = null,
    val errorMessage: String? = null
)

class GroupWriteViewModel(
    private val groupRepository: GroupRepository,
    private val memberRepository: MemberRepository,
    private val campusRepository: CampusRepository,
    private val groupKind: GroupKind
) : ViewModel() {

    private val _uiState = MutableStateFlow(GroupWriteUiState())
    val uiState: StateFlow<GroupWriteUiState> = _uiState.asStateFlow()

    init {
        loadCampusId()
    }

    fun submit(info: GroupCreateInfo) = viewModelScope.launch {
        _uiState.update { it.copy(isSubmitting = true, errorMessage = null) }
        val result = if (groupKind == GroupKind.STUDY) {
            groupRepository.createStudy(info)
        } else {
            groupRepository.createTeam(info)
        }
        result.onSuccess {
            _uiState.update { it.copy(isSubmitting = false, isSuccess = true) }
        }.onFailure { e ->
            _uiState.update { it.copy(isSubmitting = false, errorMessage = e.message) }
        }
    }

    fun update(groupId: Long, info: GroupUpdateInfo) = viewModelScope.launch {
        _uiState.update { it.copy(isSubmitting = true, errorMessage = null) }
        val result = if (groupKind == GroupKind.STUDY) {
            groupRepository.updateStudy(groupId, info)
        } else {
            groupRepository.updateTeam(groupId, info)
        }
        result.onSuccess {
            _uiState.update { it.copy(isSubmitting = false, isSuccess = true) }
        }.onFailure { e ->
            _uiState.update { it.copy(isSubmitting = false, errorMessage = e.message) }
        }
    }

    fun resetResult() {
        _uiState.update { it.copy(isSuccess = false, errorMessage = null) }
    }

    private fun loadCampusId() = viewModelScope.launch {
        val campusName = memberRepository.getMyPage().getOrNull()?.user?.campus?.trim()
        if (campusName.isNullOrBlank()) {
            _uiState.update { it.copy(campusId = null) }
            return@launch
        }
        val campuses = campusRepository.getCampuses().getOrNull().orEmpty()
        val normalized = normalizeCampusName(campusName)
        val campusId = campuses.firstOrNull { campus ->
            val campusNormalized = normalizeCampusName(campus.name)
            campusNormalized == normalized ||
                campusNormalized.contains(normalized) ||
                normalized.contains(campusNormalized)
        }?.id?.toLong()
        _uiState.update { it.copy(campusId = campusId) }
    }

    private fun normalizeCampusName(name: String): String {
        return name.replace("캠퍼스", "").replace(" ", "").lowercase()
    }
}
